server/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java

```diff
@@ -146,7 +146,20 @@ public final void run() throws IOException {
         if (shardsIts.size() > 0) {
             int maxConcurrentShardRequests = Math.min(this.maxConcurrentShardRequests, shardsIts.size());
             final boolean success = shardExecutionIndex.compareAndSet(0, maxConcurrentShardRequests);
-            assert success;
+            assert success;            
+            assert request.allowPartialSearchResults() != null : "SearchRequest missing setting for allowPartialSearchResults";
+            if (request.allowPartialSearchResults() == false) {
+                // Fail-fast verification of all shards being available
+                for (int index = 0; index < shardsIts.size(); index++) {
+                    final SearchShardIterator shardRoutings = shardsIts.get(index);
+                    if (shardRoutings.size() == 0) {
+                        //Status red - shard is missing all copies and would produce partial results for an index search
+                        final String msg = "Search rejected due to missing shard "+ shardRoutings.shardId() +
+                                ". Consider using `allow_partial_search_results` setting to bypass this error.";
+                        throw new SearchPhaseExecutionException(getName(), msg, null, ShardSearchFailure.EMPTY_ARRAY);
```

Comment:
I wonder if we should gather the missing shards and report all the shards that were unavailable rather than just the first?